framework/utils.py

```diff
@@ -148,3 +148,8 @@ def get_banner_time(timestamp):
     return None
   ts = datetime.datetime(*timestamp, tzinfo=_UTC)
   return calendar.timegm(ts.timetuple())
+
+
+def dedupe(list_with_duplicates):
+  """Return a list without duplicates, in the original order."""
+  return list(dict.fromkeys(list_with_duplicates))
```

Comment:
I started to suggest using a set for this, but then saw the filtering needs to be stable. Sets technically aren't guaranteed to be stable, and [dictionaries are as of 3.7](https://www.w3schools.com/python/python_dictionaries.asp#:~:text=Ordered%20or%20Unordered%3F-,As%20of%20Python%20version%203.7%2C%20dictionaries%20are%20ordered.).